components/quick_input/quick_input_textbox.tsx

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+import React from 'react';
+
+import AutosizeTextarea from 'components/autosize_textarea';
+
+import {QuickInputState} from './quick_input_state';
+
+interface IQInputTextboxProps{
+
+    // text box value attribute, this is old API. New api migrate this props to quickInputState
+    value: string;
+
+    // initial element to render QuickInputTextbox
+    inputComponent: any,
+
+    /**
+     * Input event callback, for old API(using ref)
+     */
+    onChange?: (e: React.ChangeEvent<HTMLInputElement | any>) => void;
+    onInput?: (e: React.ChangeEvent<HTMLInputElement | any>) => void;
+    onCompositionUpdate?: (e: React.CompositionEvent<HTMLInputElement | any>) => void;
+    onSelect?: (e: React.SyntheticEvent<HTMLInputElement, any>) => void;
+
+    /**
+     * Input event callback,for new API (using QuickInputState class to control)
+     */
+    quickInputState?: QuickInputState;
+    onTextboxChange?: (qInputState: QuickInputState) => void;
+    onTextboxCompositionUpdate?: (
+        qInputState: QuickInputState, composingData: string | null | undefined
+    ) => void;
+    onTextboxSelect?: (qInputState: QuickInputState) => void;
+}
+
+const QuickInputTextbox = React.forwardRef((props: IQInputTextboxProps, ref) => {
+    const defaultValueRef = React.useRef<string>(props.value);
+    const {
+        value, inputComponent,
+        onSelect, onCompositionUpdate, onChange, // old callbacks
+        // new callbacks
+        quickInputState, onTextboxChange, onTextboxSelect, onTextboxCompositionUpdate,
+        ...remainedProps
+    } = props;
+
+    if (inputComponent !== AutosizeTextarea) {
+        Reflect.deleteProperty(remainedProps, 'onHeightChange');
+    }
+
+    return React.createElement(
+        props.inputComponent || 'input',
+        {
+            ...remainedProps,
+            defaultValue: defaultValueRef.current,
+            ref,
+
+            // add value as props (controlled components)
+            value: quickInputState?.getValue() || value,
+
+            // override default onChange event
+            onChange: (e: React.ChangeEvent<HTMLInputElement | any>) => {
```

Comment:
You could pass the whole event target and filter the props needed in the quickinput state.replace

Reply:
I don't think that is recommendable. The point of event handler is to handle new state changed of the component (for`QuickInput` is the `QuickInputState`, or for `input` is the `React.ChangeEvent`). Passing the whole event target to a component controlled by `QuickInputState` is not really logical, I think.

Reply:
Okay, I think I should revert `QuickInput` to its original API, since `SuggestionBox` is on its way of getting rid of `QuickInput` as its child. Refactoring `QuickInput` should be on a separate PR.